hazelcast/discovery.py

```diff
@@ -0,0 +1,169 @@
+        """
+        try:
+            return list(self.cloud_discovery.discover_nodes().keys())
+        except Exception as ex:
+            self.logger.warning("Failed to load addresses from hazelcast.cloud: {}".format(ex.args[0]))
+        return []
+
+
+class HazelcastCloudAddressTranslator(object):
+    """
+    Resolves private IP addresses of Hazelcast.cloud service.
+    """
+    logging.basicConfig()
+    logger = logging.getLogger("HazelcastAddressTranslator")
+
+    def __init__(self, host, url, connection_timeout):
+        self.cloud_discovery = HazelcastCloudDiscovery(host, url, connection_timeout)
+        self._private_to_public = dict()
+
+    def translate(self, address):
+        """
+        Translates the given address to another address specific to network or service.
+        :param address: (:class:`~hazelcast.core.Address`), private address to be translated
+        :return: (:class:`~hazelcast.core.Address`), new address if given address is known, otherwise returns null
+        """
+        if address is None:
+            return None
+
+        public_address = self._private_to_public.get(address)
+        if public_address:
+            return public_address
+
+        self.refresh()
+
+        return self._private_to_public.get(address)
+
+    def refresh(self):
+        """
+        Refreshes the internal lookup table if necessary.
+        """
+        try:
+            self._private_to_public = self.cloud_discovery.discover_nodes()
+        except Exception as ex:
+            self.logger.warning("Failed to load addresses from hazelcast.cloud: {}".format(ex.args[0]))
+
+
+class HazelcastCloudDiscovery(object):
+    """
+    Discovery service that discover nodes via hazelcast.cloud
+    https://coordinator.hazelcast.cloud/cluster/discovery?token=<TOKEN>
+    """
+    _CLOUD_URL_PATH = "/cluster/discovery?token="
+    _CERTIFICATE_URL_PATH = "/cluster/certificate?token="
+    _PRIVATE_ADDRESS_PROPERTY = "private-address"
+    _PUBLIC_ADDRESS_PROPERTY = "public-address"
+
+    CLOUD_URL_BASE_PROPERTY = ClientProperty("hazelcast.client.cloud.url", "https://coordinator.hazelcast.cloud")
+    """
+    Internal client property to change base url of cloud discovery endpoint.
+    Used for testing cloud discovery.
+    """
+
+    def __init__(self, host, url, connection_timeout):
+        self._host = host
+        self._url = url
+        self._connection_timeout = connection_timeout
+        # Default context operates only on TLSv1+, checks certificates and hostname
+        self._ctx = ssl.create_default_context()
+
+    def discover_nodes(self):
+        """
+        Discovers nodes from Hazelcast.cloud.
+        :return: (dict), dictionary that maps private addresses to public addresses.
+        """
+        try:
+            return self._call_service()
+        except Exception as ex:
+            raise ex
+
+    def get_certificate(self, discovery_token):
```

Comment:
I think this method should be named `download_certificate`. I think it would be better to give the complete URL from which it will try to download the certificate.